src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs

```diff
@@ -4551,19 +4549,84 @@ private void EmitCallCharInClass(string charClass, bool caseInsensitive, LocalBu
                 return;
             }
 
-            // Next, special-case ASCII inputs.  If the character class contains only ASCII inputs, then we
-            // can satisfy the entire operation via a small lookup table, e.g.
-            //     ch < 128 && lookup(ch)
-            // If the character class contains values outside of the ASCII range, we can still optimize for
-            // ASCII inputs, using the table for values < 128, and falling back to calling CharInClass
-            // for anything outside of the ASCII range, e.g.
-            //     if (ch < 128) lookup(ch)
-            //     else ...
-            // Either way, we need to generate the lookup table for the ASCII range.
-            // We use a const string instead of a byte[] / static data property because
-            // it lets IL emit handle all the gory details for us.  It also is ok from an
-            // endianness perspective because the compilation happens on the same machine
-            // that runs the compiled code.
+            // Store the input character to a local so we can read it multiple times.
+            Stloc(tempLocal);
+
+            // Analyze the character set more to determine what code to generate.
+            RegexCharClass.CharClassAnalysisResults analysis = RegexCharClass.Analyze(charClass);
+
+            // Helper method that emits a call to RegexRunner.CharInClass(ch{.ToLowerInvariant()}, charClass)
+            void EmitCharInClass()
+            {
+                Ldloc(tempLocal);
+                if (invariant)
+                {
+                    CallToLower();
+                }
+                Ldstr(charClass);
+                Call(s_charInClassMethod);
+            }
+
+            Label doneLabel = DefineLabel();
+            Label comparisonLabel = DefineLabel();
+
+            if (!invariant) // if we're being asked to do a case insensitive, invariant comparison, use the lookup table
```

Comment:
For ASCII-only ranges, might be worth trying masking with `0x20` instead of the lookup table.

Reply:
Can you elaborate?  How would it avoid the lookup table?